### PR TITLE
Implement Kotlin Binary Compatibility Validator

### PR DESCRIPTION
**Pull request by Toumouz**
## Description

Add plugin to root project and filter out Sample and Test-utils Sub-Modules 
First apiDump completed and validated
Validated Public API and made sure CI is triggering apiCheck

## Motivation and Context

To help prevent introduction of breaking changes to Tricot API.
We can integrate this Kotlin Binary-Compatibility-Validator to compare and validate the binary code changes before every commit.
The CI gradle check will verify the api binary and fail if changes were made.

With this setup, we can now keep track of intentional changes and document them and we can now catch non-expected changes to the api and fix them before merging

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -47,3 +47,6 @@ jobs:
 
       - name: Build with Gradle
         run: ./gradlew check --parallel
+
+      - name: API check
```
Comment (npresseault) [inline .github/workflows/ci.yml:51]: Correct me if I'm wrong but I think `check` already depends on `apiCheck`, so there would be no need to call it explicitely.

File: build.gradle.kts [hunk 2]
```diff
@@ -41,6 +42,27 @@ subprojects {
...
+/**
+ * Kotlin binary compatibility validator configuration
+ * https://github.com/Kotlin/binary-compatibility-validator#setup
+ */
+apiValidation {
+    /* Packages that are excluded from public API dumps even if they contain public API. */
+    // ignoredPackages.add("kotlinx.coroutines.internal")
+
+    /* Sub-projects that are excluded from API validation */
+    ignoredProjects.addAll(listOf("sample", "android", "common", "test-utils"))
```
Comment (npresseault) [inline build.gradle.kts:54]: We could include test-utils, it's a collection of utils that can be / is used from outside trikot.

File: build.gradle.kts [hunk 3]
```diff
@@ -41,6 +42,27 @@ subprojects {
...
+
+    /* Classes (fully qualified) that are excluded from public API dumps even if they contain public API. */
+    // ignoredClasses.addAll(listOf("", ""))
```
Comment (npresseault) [inline build.gradle.kts:57]: cleanup comments

File: build.gradle.kts [hunk 4]
```diff
@@ -41,6 +42,27 @@ subprojects {
...
+
+    /* Set of annotations that exclude API from being public. */
+    // nonPublicMarkers.add("")
+
+    /* Flag to programmatically disable compatibility validator */
+    validationDisabled = false
```
Comment (npresseault) [inline build.gradle.kts:63]: what's the default value on this?

File: trikot-streams/streams/api/android/streams.api [hunk 5]
```diff
@@ -0,0 +1,929 @@
+public final class ScanWithSeedProcessor : com/mirego/trikot/streams/reactive/processors/AbstractProcessor {
```
Comment (npresseault) [inline trikot-streams/streams/api/android/streams.api:1]: processors could be internal

File: trikot-streams/streams/api/android/streams.api [hunk 6]
```diff
@@ -0,0 +1,929 @@
...
+	public abstract fun createSubscription (Lorg/reactivestreams/Subscriber;)Lcom/mirego/trikot/streams/reactive/processors/ProcessorSubscription;
+	public final fun getParentPublisher ()Lorg/reactivestreams/Publisher;
+	public fun onComplete ()V
+	public fun onError (Ljava/lang/Throwable;)V
+	public fun onNext (Ljava/lang/Object;)V
+	public fun onSubscribe (Lorg/reactivestreams/Subscription;)V
+	public fun subscribe (Lorg/reactivestreams/Subscriber;)V
+}
+
+public final class com/mirego/trikot/streams/reactive/processors/CombineLatestProcessor : com/mirego/trikot/streams/reactive/processors/AbstractProcessor {
```
Comment (npresseault) [inline trikot-streams/streams/api/android/streams.api:387]: processors could be internal, same comment for all the other ones.

File: trikot-viewmodels-declarative-compiler/viewmodels-declarative-compiler-core/api/viewmodels-declarative-compiler-core.api [hunk 7]
```diff
@@ -0,0 +1,43 @@
+public abstract class com/mirego/trikot/viewmodels/declarative/compiler/BaseBuilderProcessor : com/google/devtools/ksp/processing/SymbolProcessor {
```
Comment (npresseault) [inline trikot-viewmodels-declarative-compiler/viewmodels-declarative-compiler-core/api/viewmodels-declarative-compiler-core.api:1]: We can ignore this module.

File: trikot-viewmodels-declarative-compiler/viewmodels-declarative-compiler-flow/api/viewmodels-declarative-compiler-flow.api [hunk 8]
```diff
@@ -0,0 +1,22 @@
+public final class com/mirego/trikot/viewmodels/declarative/compiler/BuilderProcessor : com/mirego/trikot/viewmodels/declarative/compiler/BaseBuilderProcessor {
```
Comment (npresseault) [inline trikot-viewmodels-declarative-compiler/viewmodels-declarative-compiler-flow/api/viewmodels-declarative-compiler-flow.api:1]: We can ignore this module.

File: trikot-viewmodels-declarative-compiler/viewmodels-declarative-compiler-streams/api/viewmodels-declarative-compiler-streams.api [hunk 9]
```diff
@@ -0,0 +1,22 @@
+public final class com/mirego/trikot/viewmodels/declarative/compiler/BuilderProcessor : com/mirego/trikot/viewmodels/declarative/compiler/BaseBuilderProcessor {
```
Comment (npresseault) [inline trikot-viewmodels-declarative-compiler/viewmodels-declarative-compiler-streams/api/viewmodels-declarative-compiler-streams.api:1]: We can ignore this module.

**Review by npresseault — APPROVED** [review]

File: build.gradle.kts (hunk 4 above)
Comment (Toumouz) [reply build.gradle.kts:69]: It is false, will remove

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Toumouz) [reply .github/workflows/ci.yml:51]: It does!

**Pull request merged**